File: 16/python/main.py
def to_int(bits):
    return int(''.join(str(x) for x in bits), 2)

bits = {
'0': '0000',
'1': '0001',
'2': '0010',
'3': '0011',
'4': '0100',
'5': '0101',
'6': '0110',
'7': '0111',
'8': '1000',
'9': '1001',
'A': '1010',
'B': '1011',
'C': '1100',
'D': '1101',
'E': '1110',
'F': '1111',
}

def to_bits(x):
    return [int(a) for  a in bits[x]]


def process_op(type, sub):
    print("In here", type, sub)
    if type == 0:
        return sum(sub)
    if type == 1:
        prod = 1
        for x in sub:
            prod *= x
        return prod
    if type == 2:
        return min(sub)
    if type == 3:
        return max(sub)
    if type == 5:
        return 1 if sub[0] > sub[1] else 0
    
    if type == 6:
        return 1 if sub[0] < sub[1] else 0
    
    if type == 7:
        return 1 if sub[0] == sub[1] else 0

total_version = 0
def process_packet(bits):
    print(bits)
    end = 0
    version = to_int(bits[0:3])
    print("version", version)
    global total_version
    total_version += version
    # print(total_version)
    packet_type = to_int(bits[3:6])
    print("packet type", packet_type)
    if (packet_type != 4):
        sub = bits[6]
        if sub == 0:
            
            length = to_int(bits[7:7+15])
            # print(length)
            x = 7+15
            end = 7+15+length
            sub_vals = []
            while x < end:
                a,b = process_packet(bits[x:])
                x += a
                sub_vals.append(b)
                
            return (7+15+length, process_op(packet_type, sub_vals))
        else:
            count = to_int(bits[7:7+11])
            x = 7+11
            sub_vals = []
            for i in range(0, count):
                a,b = process_packet(bits[x:])
                x += a
                sub_vals.append(b)

            return (x, process_op(packet_type, sub_vals))
    else:
        x = 6
        val = []
        # print(bits)
        while bits[x] != 0:
            val += bits[x+1:x+5]
            x += 5
        val += bits[x+1:x+5]
        # print("literal", to_int(val))
        return (x + 5, to_int(val))
    # print(version)

def solution(lines):
    v = []
    for x in lines[0]:
        v += to_bits(x)
    # v = [to_bits(x) for x in lines[0]]
    print(process_packet(v))
    # print(total_version)

with open('input.txt') as f:
    lines = f.readlines()
    solution(lines)

File: 16/haskell/main.hs
import Data.Char (digitToInt)
import Data.List
import Debug.Trace
import Numeric (readHex)
import System.IO
import Text.Printf

type Bit = Int

data Packet
  = LiteralPacket Int Int
  | OperatorPacket Int Int [Packet]
  deriving (Show)

-- Generic bit twiddling stuff...
toDec :: String -> Int
toDec = foldl' (\acc x -> acc * 2 + digitToInt x) 0

bitsToInt :: [Bit] -> Int
bitsToInt bits = toDec $ concatMap show bits

charToBits :: Char -> [Bit]
charToBits c = case readHex [c] of
  (x, _) : _ -> map digitToInt $ printf "%04b" (x :: Int)
  _ -> error "Failed to parse character"

toBitArray :: String -> [Bit]
toBitArray = concatMap charToBits

getLiteralBits :: [Bit] -> ([Bit], [Bit])
getLiteralBits (0 : xs) = splitAt 4 xs
getLiteralBits (1 : xs) = (take 4 xs ++ content, rest)
  where
    (content, rest) = getLiteralBits (drop 4 xs)
getLiteralBits _ = error "Unexpected literal bits"

getLiteralValue :: [Bit] -> (Int, [Bit])
getLiteralValue bits = (bitsToInt val, rest)
  where
    (val, rest) = getLiteralBits bits

getSubPacketLengthHelper :: Int -> ([Packet], [Bit]) -> ([Packet], [Bit])
getSubPacketLengthHelper remaining (packets, bits) =
  if remaining == 0
    then (packets, bits)
    else getSubPacketLengthHelper (remaining - (length bits - length rest)) (packets ++ [parsed], rest)
  where
    (parsed, rest) = parsePacket bits

getSubPackets :: [Bit] -> ([Packet], [Bit])
getSubPackets (0 : bits) = getSubPacketLengthHelper subLength ([], drop 15 bits)
  where
    subLength = bitsToInt $ take 15 bits
getSubPackets (1 : bits) = foldl' helper ([], drop 11 bits) [1 .. subPacketCount]
  where
    subPacketCount = bitsToInt $ take 11 bits
    helper (packets, remainingBits) _ = (packets ++ [parsed], rest)
      where
        (parsed, rest) = parsePacket remainingBits
getSubPackets _ = error "Unexpected bits when parsing sub packets"

buildLiteralPacket :: Int -> [Bit] -> (Packet, [Bit])
buildLiteralPacket version bits = (LiteralPacket version val, rest)
  where
    (val, rest) = getLiteralValue bits

buildOperatorPacket :: Int -> Int -> [Bit] -> (Packet, [Bit])
buildOperatorPacket version opCode bits = (OperatorPacket version opCode packets, rest)
  where
    (packets, rest) = getSubPackets bits

parsePacket :: [Bit] -> (Packet, [Bit])
parsePacket bits = case packetType of
  4 -> buildLiteralPacket version $ drop 6 bits
  _ -> buildOperatorPacket version packetType $ drop 6 bits
  where
    version = bitsToInt $ take 3 bits
    packetType = bitsToInt $ take 3 $ drop 3 bits

sumVersions :: Packet -> Int
sumVersions (LiteralPacket version _) = version
sumVersions (OperatorPacket version _ packets) = version + sum (map sumVersions packets)

evaluate :: Packet -> Int
evaluate (LiteralPacket _ val) = val
evaluate (OperatorPacket _ 0 packets) = sum $ map evaluate packets
evaluate (OperatorPacket _ 1 packets) = product $ map evaluate packets
evaluate (OperatorPacket _ 2 packets) = minimum $ map evaluate packets
evaluate (OperatorPacket _ 3 packets) = maximum $ map evaluate packets
evaluate (OperatorPacket _ 5 packets) = if evaluate (head packets) > evaluate (last packets) then 1 else 0
evaluate (OperatorPacket _ 6 packets) = if evaluate (head packets) < evaluate (last packets) then 1 else 0
evaluate (OperatorPacket _ 7 packets) = if evaluate (head packets) == evaluate (last packets) then 1 else 0
evaluate _ = error "Unknown operator"

solution1 :: String -> Int
solution1 = sumVersions . fst . parsePacket . toBitArray

solution2 :: String -> Int
solution2 = evaluate . fst . parsePacket . toBitArray

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  print (solution1 contents)
  print (solution2 contents)
  hClose handle
